File: typescript/src/questions/chapter2.ts
/**
 * Not using the actual LinkedList implementation since it makes most of these questions trivial.
 * Instead, use the internal implementation node.
 * For all questions that need a singly linked list, the prev pointer is simply ignored.
 */
import { LinkedListNode } from "linked-list";
import { LinkedStack } from "stack";

/**
 * 2.1 Remove Dups: Write a function to remove duplicates from an unsorted linked list.
 * What if an extra buffer was not allowed (i.e. O(1) space is needed)?
 * Assume that the list contains numbers.
 */

/** Time: O(n). Space: O(n). */
export function deleteDuplicatesBuffer(head: LinkedListNode<number>) {
  const visited: Record<number, boolean> = {};
  let previous: LinkedListNode<number>;
  let currentNode = head;
  while (currentNode) {
    if (visited[currentNode.value]) {
      previous.next = currentNode.next;
    } else {
      visited[currentNode.value] = true;
      previous = currentNode;
    }
    currentNode = currentNode.next;
  }
}

/** Time: O(n^2). Space: O(1). */
export function deleteDuplicatesNoBuffer(head: LinkedListNode<number>) {
  let currentNode = head;
  while (currentNode) {
    let runner = currentNode;
    while (runner.next) {
      if (runner.next.value === currentNode.value) {
        runner.next = runner.next.next;
      } else {
        runner = runner.next;
      }
    }
    currentNode = currentNode.next;
  }
}

/**
 * 2.2 Return Nth to Last: Write a function to find the nth to last element of a singly linked list.
 * Assume that the size of the list is not available, since this makes the solution trivial.
 */

/** Time: O(n). Space: O(n) heap space. */
export function nthToLastRecursive(head: LinkedListNode<number>, n: number) {
  return nthToLastHelper(head, n, { value: 0 });
}

function nthToLastHelper(node: LinkedListNode<number>, n: number, tracer: { value: number }): number {
  if (!node) {
    return undefined;
  }
  const nd = nthToLastHelper(node.next, n, tracer);
  tracer.value++;
  if (tracer.value === n) {
    return node.value;
  }
  return nd;
}

/** Time: O(n). Size: O(1). */
export function nthToLastIterative(head: LinkedListNode<number>, n: number) {
  let currentNode = head;
  let scout = head;

  for (let i = 0; i < n; i++) {
    if (!scout) {
      return undefined;
    }
    scout = scout.next;
  }

  while (scout) {
    scout = scout.next;
    currentNode = currentNode.next;
  }
  return currentNode.value;
}

/**
 * 2.3 Delete Middle Node: Write a function to delete an arbitrary middle node of a
 * singly linked list, given only access to that node.
 * Assume that the given node cannot be the first or the last node.
 */

/** Time: O(1). Space: O(1). */
export function deleteMiddleNode(node: LinkedListNode<number>) {
  if (!node || !node.next) {
    throw new Error("Can only delete middle nodes");
  }
  const next = node.next;
  node.value = next.value;
  node.next = next.next;
}

/**
 * 2.4 Partition: Write a function to partition a linked list around a value x,
 * such that all nodes less than x come before all nodes great than or equal to x.
 * The relative order of the elements should be preserved.
 * The head of the partitioned list should be returned by the function.
 */

/** Time: O(n). Space: O(n). */
export function partitionList(head: LinkedListNode<number>, x: number) {
  // Three linkedlists to hold smaller, larger, and equal values.
  let smallerHead: LinkedListNode<number>, smallerTail: LinkedListNode<number>;
  let greaterHead: LinkedListNode<number>, greaterTail: LinkedListNode<number>;
  let equalHead: LinkedListNode<number>, equalTail: LinkedListNode<number>;
  let currentNode = head;

  // Add all elements to their appropriate list.
  while (currentNode) {
    if (currentNode.value === x) {
      if (!equalHead) {
        equalHead = currentNode;
        equalTail = currentNode;
      } else {
        equalTail.next = currentNode;
        equalTail = currentNode;
      }
    } else if (currentNode.value < x) {
      if (!smallerHead) {
        smallerHead = currentNode;
        smallerTail = currentNode;
      } else {
        smallerTail.next = currentNode;
        smallerTail = currentNode;
      }
    } else {
      if (!greaterTail) {
        greaterHead = currentNode;
        greaterTail = currentNode;
      } else {
        greaterTail.next = currentNode;
        greaterTail = currentNode;
      }
    }
    currentNode = currentNode.next;
  }
  if (greaterTail) {
    greaterTail.next = undefined;
  }
  // Return the concatenation of the 3 lists.
  if (!smallerHead) {
    if (!equalHead) {
      return greaterHead;
    }
    equalTail.next = greaterHead;
    return equalHead;
  }
  if (!equalHead) {
    smallerTail.next = greaterHead;
    return smallerHead;
  }
  smallerTail.next = equalHead;
  equalTail.next = greaterHead;
  return smallerHead;
}

/**
 * 2.5 Sum Lists: Given two integers represented by linked lists, where each node contains a
 * single digit, write a function that adds the two numbers into a new linked list and returns the head.
 * The linked lists store numbers in reverse order.
 */

/** Time: O(n), where n is the length of the longer number. Space: O(n) heap space. */
export function sumListsReverse(head1: LinkedListNode<number>, head2: LinkedListNode<number>) {
  return sumListsReverseHelper(head1, head2, 0);
}

function sumListsReverseHelper(node1: LinkedListNode<number>, node2: LinkedListNode<number>, carry: number) {
  if (!node1 && !node2 && carry === 0) {
    return undefined;
  }
  let value = carry;
   // Add the carry to the sum.
  if (node1) {
    value += node1.value;
  }
  if (node2) {
    value += node2.value;
  }
  // Take the ones digit of the sum.
  const result = new LinkedListNode<number>(value % 10, undefined, undefined);
  if (node1 || node2) {
    const temp = sumListsReverseHelper(
      node1 ? node1.next : undefined,
      node2 ? node2.next : undefined,
      value >= 10 ? 1 : 0
    );
    result.next = temp;
  }
  return result;
}

/**
 * 2.6 Palindrome List: Implement a function to check if a linked list a palindrome.
 */

/** Time: O(n). Space: O(n). */
export function isListPalindrome(head: LinkedListNode<string>) {
  let currentNode = head;
  let runner = head;
  const stack = new LinkedStack<string>();
  while (runner && runner.next) {
    stack.push(currentNode.value);
    currentNode = currentNode.next;
    runner = runner.next.next;
  }
  if (runner) { // List has an odd number of nodes, skip the middle ones.
    currentNode = currentNode.next;
  }
  while (currentNode) {
    if (stack.pop() !== currentNode.value) {
      return false;
    }
    currentNode = currentNode.next;
  }
  return true;
}

/**
 * 2.7 List Intersection: Given two singly linked lists of size M and N, write a function to determine
 * if the two lists intersect, and if they do, return the intersecting node.
 */

/** Time: O(M + N). Space: O(1). */
export function listIntersect(head1: LinkedListNode<number>, head2: LinkedListNode<number>) {
  if (!head1 || !head2) {
    return undefined;
  }
  const getTailAndSize = (node: LinkedListNode<number>) => {
    let currentNode = node;
    let size = 0;
    while (currentNode.next) {
      currentNode = currentNode.next;
      size++;
    }
    return { tail: currentNode, size: size + 1 };
  };
  const tailSize1 = getTailAndSize(head1);
  const tailSize2 = getTailAndSize(head2);
  if (tailSize1.tail !== tailSize2.tail) {
    return undefined;
  }
  let shorter = tailSize1.size < tailSize2.size ? head1 : head2;
  let longer = tailSize1.size < tailSize2.size ? head2 : head1;
  for (let i = 0; i < Math.abs(tailSize1.size - tailSize2.size); i++) {
    longer = longer.next;
  }
  while (shorter !== longer) {
    shorter = shorter.next;
    longer = longer.next;
  }
  return shorter;
}

/**
 * 2.8 Loop Detection: Given a singly linked list, write a function to determine if the
 * list is circular. If it is, return the node at the beginning of the loop.
 */

/** Time: O(kn), where k is the size of the loop. Space: O(1). */
export function detectLoop(head: LinkedListNode<number>) {
  let runner = head;
  let currentNode = head;

  while (runner && runner.next) {
    currentNode = currentNode.next;
    runner = runner.next.next;
    if (runner === currentNode) {
      break;
    }
  }

  if (!runner || !runner.next) {
    return undefined;
  }

  currentNode = head;
  while (currentNode !== runner) {
    currentNode = currentNode.next;
    runner = runner.next;
  }
  return runner;
}

File: go/questions/chapter2.go
package questions

import (
	"errors"
	"math"
	"strconv"

	"../structures"
)

/** See chapter2.ts for problem descriptions. */

// DeleteDuplicatesBuffer removes duplicates from a linked list using a buffer.
// Time: O(n). Space: O(n).
func DeleteDuplicatesBuffer(head *structures.LinkedListNode) {
	visited := make(map[int]bool)
	var previous *structures.LinkedListNode
	currentNode := head
	for currentNode != nil {
		if visited[currentNode.Value] {
			previous.Next = currentNode.Next
		} else {
			visited[currentNode.Value] = true
			previous = currentNode
		}
		currentNode = currentNode.Next
	}
}

// DeleteDuplicatesNoBuffer removes duplicates from a linked list without a buffer.
// Time: O(n). Space: O(1).
func DeleteDuplicatesNoBuffer(head *structures.LinkedListNode) {
	currentNode := head
	for currentNode != nil {
		runner := currentNode
		for runner.Next != nil {
			if runner.Next.Value == currentNode.Value {
				runner.Next = runner.Next.Next
			} else {
				runner = runner.Next
			}
		}
		currentNode = currentNode.Next
	}
}

// NthToLastRecursive recursively finds the nth to last element of a linked list.
// Time: O(n). Space: O(n) heap space.
func NthToLastRecursive(head *structures.LinkedListNode, n int) (int, error) {
	tracer := 0
	return nthToLastRecursiveHelper(head, n, &tracer)
}

func nthToLastRecursiveHelper(node *structures.LinkedListNode, n int, tracer *int) (int, error) {
	if node == nil {
		return 0, errors.New("Cannot find nth to last element: " + strconv.Itoa(n))
	}
	nd, err := nthToLastRecursiveHelper(node.Next, n, tracer)
	*tracer++
	if *tracer == n {
		return node.Value, nil
	}
	return nd, err
}

// NthToLastIterative iteratively finds the nth to last element of a linked list.
// Time: O(n). Space: O(1).
func NthToLastIterative(head *structures.LinkedListNode, n int) (int, error) {
	currentNode := head
	scout := head
	for i := 0; i < n; i++ {
		if scout == nil {
			return 0, errors.New("Cannot find nth to last element: " + strconv.Itoa(n))
		}
		scout = scout.Next
	}
	for scout != nil {
		scout = scout.Next
		currentNode = currentNode.Next
	}
	return currentNode.Value, nil
}

// DeleteMiddleNode deletes a node in the middle of a linked list.
// Time: O(1). Space: O(1).
func DeleteMiddleNode(node *structures.LinkedListNode) error {
	if node == nil || node.Next == nil {
		return errors.New("Can only delete middle nodes")
	}
	next := node.Next
	node.Value = next.Value
	node.Next = next.Next
	return nil
}

// PartitionList partitions a linked list around a pivot x.
// Time: O(n). Space: O(n).
func PartitionList(head *structures.LinkedListNode, x int) *structures.LinkedListNode {
	var smallerHead, smallerTail,
		greaterHead, greaterTail,
		equalHead, equalTail *structures.LinkedListNode
	currentNode := head
	for currentNode != nil {
		if currentNode.Value == x {
			if equalHead == nil {
				equalHead = currentNode
				equalTail = currentNode
			} else {
				equalTail.Next = currentNode
				equalTail = currentNode
			}
		} else if currentNode.Value < x {
			if smallerHead == nil {
				smallerHead = currentNode
				smallerTail = currentNode
			} else {
				smallerTail.Next = currentNode
				smallerTail = currentNode
			}
		} else {
			if greaterHead == nil {
				greaterHead = currentNode
				greaterTail = currentNode
			} else {
				greaterTail.Next = currentNode
				greaterTail = currentNode
			}
		}
		currentNode = currentNode.Next
	}
	if smallerHead == nil {
		if equalHead == nil {
			return greaterHead
		}
		equalTail.Next = greaterHead
		return equalHead
	}
	if equalHead == nil {
		smallerTail.Next = greaterHead
		return smallerHead
	}
	smallerTail.Next = equalHead
	equalTail.Next = greaterHead
	return smallerHead
}

// SumListsReverse sums two numbers represented in reverse by linked lists.
// Time: O(n), where n is the length of the longer number. Space: O(n) heap space.
func SumListsReverse(head1 *structures.LinkedListNode, head2 *structures.LinkedListNode) *structures.LinkedListNode {
	return sumListsReverseHelper(head1, head2, 0)
}

func sumListsReverseHelper(node1 *structures.LinkedListNode, node2 *structures.LinkedListNode, carry int) *structures.LinkedListNode {
	if node1 == nil && node2 == nil && carry == 0 {
		return nil
	}
	value := carry
	if node1 != nil {
		value += node1.Value
	}
	if node2 != nil {
		value += node2.Value
	}
	result := &structures.LinkedListNode{Value: value % 10}
	if node1 != nil || node2 != nil {
		var next1, next2 *structures.LinkedListNode
		newCarry := 0
		if node1 != nil {
			next1 = node1.Next
		}
		if node2 != nil {
			next2 = node2.Next
		}
		if value >= 10 {
			newCarry = 1
		}
		temp := sumListsReverseHelper(next1, next2, newCarry)
		result.Next = temp
	}
	return result
}

// IsListPalindrome returns true if the linked list is a palindrome.
// Time: O(n). Space: O(n).
func IsListPalindrome(head *structures.LinkedListNode) bool {
	currentNode := head
	runner := head
	stack := &structures.LinkedStack{}
	for runner != nil && runner.Next != nil {
		stack.Push(currentNode.Value)
		currentNode = currentNode.Next
		runner = runner.Next.Next
	}
	if runner != nil {
		currentNode = currentNode.Next
	}
	for currentNode != nil {
		value, _ := stack.Pop()
		if value != currentNode.Value {
			return false
		}
		currentNode = currentNode.Next
	}
	return true
}

// ListIntersect returns the intersecting node of two linked lists of size M and N if it exists.
// Time: O(M + N). Space: O(1).
func ListIntersect(head1 *structures.LinkedListNode, head2 *structures.LinkedListNode) *structures.LinkedListNode {
	if head1 == nil || head2 == nil {
		return nil
	}
	type TailAndSize struct {
		tail *structures.LinkedListNode
		size int
	}
	getTailAndSize := func(node *structures.LinkedListNode) *TailAndSize {
		currentNode := node
		size := 0
		for currentNode.Next != nil {
			currentNode = currentNode.Next
			size++
		}
		return &TailAndSize{tail: currentNode, size: size + 1}
	}
	tailSize1 := getTailAndSize(head1)
	tailSize2 := getTailAndSize(head2)
	if tailSize1.tail != tailSize2.tail {
		return nil
	}
	var shorter, longer *structures.LinkedListNode
	if tailSize1.size < tailSize2.size {
		shorter = head1
		longer = head2
	} else {
		shorter = head2
		longer = head1
	}
	for i := 0; i < int(math.Abs(float64(tailSize1.size)-float64(tailSize2.size))); i++ {
		longer = longer.Next
	}
	for shorter != longer {
		shorter = shorter.Next
		longer = longer.Next
	}
	return shorter
}

// DetectLoop returns the start of a cycle in a linked list if it exists.
// Time: O(kn), where k is the size of the loop. Space: O(1).
func DetectLoop(head *structures.LinkedListNode) *structures.LinkedListNode {
	runner := head
	currentNode := head

	for runner != nil && runner.Next != nil {
		currentNode = currentNode.Next
		runner = runner.Next.Next
		if runner == currentNode {
			break
		}
	}

	if runner == nil || runner.Next == nil {
		return nil
	}

	currentNode = head
	for currentNode != runner {
		currentNode = currentNode.Next
		runner = runner.Next
	}
	return runner
}
